### ISSUE TITLE
Sonoff Pow Energy total

### ISSUE THREAD

**Issue by gajotnt** [arendst/Sonoff-Tasmota#1202]
Hello there, is there anyway to set the "Energy Total" to a specific number? So that instead of beginning from zero it will match the Grid operator counter value

Thanks

**Comment by davidelang** [comment]
no, do that sort of thing in your home automation software.

**Comment by PeteBa** [comment]
A slightly different use-case.

I think "Energy Total" should reflect, to the best of our ability, the total energy used by the upstream device rather than what that particular sonoff may have measured. For example, I had to replace a faulty sonoff and wanted the new one to "continue on" from where the old one left off. I dont think a hardware fault should require a home automation fix and better to correct the problem at source and restore the integrity of the "Energy Total" stream.

**Comment by davidelang** [comment]
The energy total will be reset to zero if the sonoff reboots during the day. 
It's not as reliable as you think.

**Comment by PeteBa** [comment]
@davidelang , Curious ... just  tested this on three sonoffs with both a restart and a power cycle. In all three cases the Energy Total was retained and confirmed via the device's web page. I'm on a fairly old version 5.2.4 but looking at the code suggests kWhTotal is stored in RTC memory. What are you seeing ?

**Comment by PeteBa** [comment]
@davidelang , an update...

How very interesting. While I am seeing continuity of kWhTotal across a soft restart I am NOT seeing that during a hard power cycle. It is not that the counter resets to zero but rather it reverts to some previously saved value. I believe that this might be the value from the previous midnight from looking at the code in `hlw_200ms()`. So you could lose up to a day's worth of usage even with SaveDate set.

Looking deeper into `StateLoop()`, it seems that `Settings.kWhTotal` is indeed regularly backed-up when SaveData is set but it is not consistently refreshed from `RtcSettings.hlw_kWhTotal` before doing so.

@arendst , is this a defect or is there a reason that `HlwSaveState()` is not called before `SettingsSave()` when `save_data_counter <= 0` as it is done before a restart? Thoughts?

**Comment by arendst** [comment]
@PeteBa To spare flash writes HlwSaveState is only called in controlled situations. To have some kind of Total Energy I save it regularly in RTC (RAM) and at the end of the day in flash too. On hard power off you will loose the amount used since last midnight save.

**Comment by PeteBa** [comment]
@arendst , yeah, that corresponds to what I am seeing here and I understand the compromise between flash wear-and-tear and keeping the Energy Total stream resilient. I do have two suggestions and happy to raise as a PR if you think they have merit:

1) Add both HlwSaveState()/CounterSaveState() to CMND_SAVEDATA handling but leave the StateLoop() as-is. This would allow the user to "lock-in" all persistent data but still save on flash wear-and-tear. Simple use-case would be to run "SaveData" before powering down to move a device to a new location.

2) Allow the user to "true-up" the Energy Total counter if it gets out of sync (e.g. scenario as per OP post and my hardware replacement scenario) with a new EnergyTotal command (or an optional parameter to the existing EnergyReset commands).

I have both of these running locally and gone through a few test cases but will let it soak for a bit.

**Comment by arendst** [comment]
I will replace the current EnergyReset functionality by this:
```
  else if ((CMND_ENERGYRESET == command_code) && (index > 0) && (index <= 3)) {
/*
    if ((payload >= 1) && (payload <= 3)) {
      switch (payload) {
      case 1:
        hlw_kWhtoday = 0;
        RtcSettings.hlw_kWhtoday = 0;
        Settings.hlw_kWhtoday = 0;
        break;
      case 2:
        Settings.hlw_kWhyesterday = 0;
        break;
      case 3:
        RtcSettings.hlw_kWhtotal = 0;
        Settings.hlw_kWhtotal = RtcSettings.hlw_kWhtotal;
        break;
      }
    }
*/
    char *p;
    long lnum = strtol(dataBuf, &p, 10);
    if (p != dataBuf) {
      lnum *= 100;
      switch (index) {
      case 1:
        hlw_kWhtoday = lnum *1000;
        RtcSettings.hlw_kWhtoday = hlw_kWhtoday;
        Settings.hlw_kWhtoday = hlw_kWhtoday;
        break;
      case 2:
        Settings.hlw_kWhyesterday = lnum *1000;
        break;
      case 3:
        RtcSettings.hlw_kWhtotal = lnum;
        Settings.hlw_kWhtotal = RtcSettings.hlw_kWhtotal;
        break;
      }
    }
    char syesterday_energy[10];
    char stoday_energy[10];
    char stotal_energy[10];
    dtostrfd((float)Settings.hlw_kWhyesterday / 100000000, Settings.flag2.energy_resolution, syesterday_energy);
    dtostrfd((float)RtcSettings.hlw_kWhtoday / 100000000, Settings.flag2.energy_resolution, stoday_energy);
    dtostrfd((float)(RtcSettings.hlw_kWhtotal + (hlw_kWhtoday / 1000)) / 100000, Settings.flag2.energy_resolution, stotal_energy);
    snprintf_P(mqtt_data, sizeof(mqtt_data), PSTR("{\"%s\":{\"" D_TOTAL "\":%s,\"" D_YESTERDAY "\":%s,\"" D_TODAY "\":%s}}"),
      command, stotal_energy, syesterday_energy, stoday_energy);
    status_flag = 1;
  }
```
Allowing the user to use commands ``EnergyReset1 0..42500``, ``EnergyReset2 0-42500`` and ``EnergyReset3 0..42500000`` to (Re)set Energy Today, Yesterday or Total in Wh

**Comment by PeteBa** [comment]
@arendst , that's fantastic.

I did a couple of different things but not sure that they are any better to be honest and I really like the index trick! I used `strtof()` to allow the user to enter kWh matching the UI format and precision. Also, I figured people would expect `EnergyReset3` to correspond to the total in the UI so I had to subtract out kWhtoday before storing. Given you are providing all three variants it should be pretty evident how they fit together.

Curious what you thought about the other point on saving counters/totals at both midnight and on initial execution of SaveData command ?
